pkg/virt-controller/services/launchercommandrenderer.go

```diff
@@ -0,0 +1,142 @@
+	isNonRoot                bool
+	gracePeriodSeconds       int
+	numberOfHookSidecars     int
+	keepAfterFailure         bool
+	namespace                string
+	shouldSimulateCrash      bool
+	virtLauncherStaticConfig VirtLauncherStaticConfig
+	vmUID                    string
+}
+
+type VirtLauncherStaticConfig struct {
+	containerDiskDir string
+	ephemeralDiskDir string
+	launcherTimeout  int
+	ovmfPath         string
+	virtShareDir     string
+}
+
+type VirtLauncherCommandRendererOption func(renderer *VirtLauncherCommandRenderer)
+
+func NewVirtLauncherCommandRenderer(
+	vmUID string,
+	domainName string,
+	gracePeriodSeconds int,
+	hookSidecarListLen int,
+	vmNamespace string,
+	virtLauncherStaticConfig VirtLauncherStaticConfig,
+	opts ...VirtLauncherCommandRendererOption,
+) *VirtLauncherCommandRenderer {
+
+	return applyOptionsToRenderer(
+		&VirtLauncherCommandRenderer{
+			domainName:               domainName,
+			gracePeriodSeconds:       gracePeriodSeconds,
+			numberOfHookSidecars:     hookSidecarListLen,
+			namespace:                vmNamespace,
+			virtLauncherStaticConfig: virtLauncherStaticConfig,
+			vmUID:                    vmUID,
+		},
+		opts,
+	)
+}
+
+func applyOptionsToRenderer(virtLauncherCmdRenderer *VirtLauncherCommandRenderer, opts []VirtLauncherCommandRendererOption) *VirtLauncherCommandRenderer {
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+		opt(virtLauncherCmdRenderer)
+	}
+	return virtLauncherCmdRenderer
+}
+
+func NewDoppelgangerPodRender(opts ...VirtLauncherCommandRendererOption) *VirtLauncherCommandRenderer {
+	return applyOptionsToRenderer(
+		&VirtLauncherCommandRenderer{isDoppelgangerPod: true},
+		opts,
+	)
+}
+
+func (vlcr *VirtLauncherCommandRenderer) Render() []string {
+	var command []string
+	if vlcr.isDoppelgangerPod {
+		command = []string{"/bin/bash",
+			"-c",
+			"echo", "bound PVCs"}
+	} else {
+		const launcherExecBinary = "/usr/bin/virt-launcher-monitor"
+		command = []string{
+			launcherExecBinary,
+			"--qemu-timeout", generateQemuTimeoutWithJitter(vlcr.virtLauncherStaticConfig.launcherTimeout),
+			"--name", vlcr.domainName,
+			"--uid", vlcr.vmUID,
+			"--namespace", vlcr.namespace,
+			"--kubevirt-share-dir", vlcr.virtLauncherStaticConfig.virtShareDir,
+			"--ephemeral-disk-dir", vlcr.virtLauncherStaticConfig.ephemeralDiskDir,
+			"--container-disk-dir", vlcr.virtLauncherStaticConfig.containerDiskDir,
+			"--grace-period-seconds", strconv.Itoa(vlcr.gracePeriodSeconds),
+			"--hook-sidecars", strconv.Itoa(vlcr.numberOfHookSidecars),
+			"--ovmf-path", vlcr.virtLauncherStaticConfig.ovmfPath,
+		}
+		if vlcr.isNonRoot {
+			command = append(command, "--run-as-nonroot")
+		}
+
+		if vlcr.customDebugFilters != "" {
+			command = append(command, "--libvirt-log-filters", vlcr.customDebugFilters)
+		}
+	}
+
+	if vlcr.allowEmulation {
+		command = append(command, "--allow-emulation")
+	}
+	if vlcr.keepAfterFailure {
+		command = append(command, "--keep-after-failure")
```

Comment:
does this flag apply for the doppelganger pod?

Reply:
Yes, but I am just preserving the existing behavior... I have no clue if it actually should. 

... honestly speaking, I don't even know the purpose of this pod.

Reply:
The doppelganger pod is used for WFFC, and I don't think these options take any effect on the pod. Actually, thinking more about it, even the flag `--allow-emulation` isn't simply a flag needed by the `virt-launcher` binary. In the case of the doppelganger pod, we are running `/bin/bash` and not `virt-launcher` binary. I'm wondering if `/bin/bash -c echo "bound PVCs"  --allow-emulation` won't fail.

Reply:
```bash
/bin/bash -c echo "bound PVCs" --allow-emulation
$ echo $?
0
```
Seems it doesn't.

Never the less, I feel that's besides the scope of this PR (fixing whatever logic on this doppelganger pod).

I would prefer (mostly motivated by my ignorance ...) if someone that actually understands what that that pod is, what it is used for, and how it should behave to tackle that on a follow-up PR.

Reply:
> > The doppelganger pod is used for WFFC
> 
> What is WFFC ? cold_sweat

WFFC stays for WaitForFirstConsumer.

Reply:
> ```shell
> /bin/bash -c echo "bound PVCs" --allow-emulation
> $ echo $?
> 0
> ```
> 
> Seems it doesn't.
> 
> Never the less, I feel that's besides the scope of this PR (fixing whatever logic on this doppelganger pod).
> 
> I would prefer (mostly motivated by my ignorance ...) if someone that actually understands what that that pod is, what it is used for, and how it should behave to tackle that on a follow-up PR.

tbh, even if it doesn't fail, I'm not a big fan of this weird construction of the command. @iholder-redhat wdyt? Should we solve it in a following PR or clarify it here?

Reply:
I'd simply set the options when it isn't a doppelganger pod. In this way:
```golang
func (vlcr *VirtLauncherCommandRenderer) Render() []string {
	if vlcr.isDoppelgangerPod {
		return []string{"/bin/bash",
			"-c",
			"echo", "bound PVCs"}
	}
	var command []string
	const launcherExecBinary = "/usr/bin/virt-launcher-monitor"
	command = []string{
		launcherExecBinary,
		"--qemu-timeout", generateQemuTimeoutWithJitter(vlcr.virtLauncherStaticConfig.launcherTimeout),
		"--name", vlcr.domainName,
		"--uid", vlcr.vmUID,
		"--namespace", vlcr.namespace,
		"--kubevirt-share-dir", vlcr.virtLauncherStaticConfig.virtShareDir,
		"--ephemeral-disk-dir", vlcr.virtLauncherStaticConfig.ephemeralDiskDir,
		"--container-disk-dir", vlcr.virtLauncherStaticConfig.containerDiskDir,
		"--grace-period-seconds", strconv.Itoa(vlcr.gracePeriodSeconds),
		"--hook-sidecars", strconv.Itoa(vlcr.numberOfHookSidecars),
		"--ovmf-path", vlcr.virtLauncherStaticConfig.ovmfPath,
	}
	if vlcr.isNonRoot {
		command = append(command, "--run-as-nonroot")
	}

	if vlcr.customDebugFilters != "" {
		command = append(command, "--libvirt-log-filters", vlcr.customDebugFilters)
	}

	if vlcr.allowEmulation {
		command = append(command, "--allow-emulation")
	}
	if vlcr.keepAfterFailure {
		command = append(command, "--keep-after-failure")
	}
	if vlcr.shouldSimulateCrash {
		command = append(command, "--simulate-crash")
	}
	return command
}
```

Reply:
Thanks for suggesting; I can comply.

Reply:
But I'd like to point out that this is not the current behavior - currently, the doppelganger pod can have the `--allow-emulation`, `--keep-after-failure`, and `--simulate-crash` flags. 

... why what that means is too much for me to comprehend.

Reply:
@alicefr thanks for educating us and catching that!

My opinion: what you're saying makes a lot of sense. If @maiqueb agrees, I'd vote to pick up your suggested changes and make sure tests pass. If nothing goes wrong, there's no reason not to catch "two birds with one stone" in this PR. However, if we start running into trouble, then I would leave it to follow-up PRs since essentially @maiqueb is right - this PR shouldn't be blocked because of this as it's a refactoring PR, not a PR focused on fixing the doppelganger pod behavior.

WDYT?

Reply:
Well, do you at least agree with me that an entrypoint with `/bin/bash -c echo "bound PVCs" --allow-emulation` doesn't make sense? I've pointed out also because you have also added it as cases to test in the unit tests. 
If we can correct a wrong way to construct the entrypoint, I just see it as a positive thing :).

Reply:
@alicefr Yes, as said above, I fully agree! It makes no sense to provide these flags to the bash binary and it's pretty much obviously a mistake. All I tried to say is that I expect this change to smoothly pass all of our tests (mainly to convince @maiqueb it's not big deal to adopt your changed :D). If, for some reason, tests would fail and we'd find out this would demand much more effort to achieve - which I don't expect happening - than it's fair IMHO to achieve that in follow-up PRs.

Reply:
I agree that entrypoint doesn't make any sense. Might have been careless of me, but I just wanted to document the current behavior. 

I'll incorporate your suggestion.